src/test/coinsviewcacheasync_tests.cpp

```diff
@@ -0,0 +1,187 @@
+            auto outpoint{in.prevout};
+            Coin coin{};
+            if (!spent) coin.out.nValue = 1;
+            BOOST_CHECK(spent ? coin.IsSpent() : !coin.IsSpent());
+            cache.EmplaceCoinInternalDANGER(std::move(outpoint), std::move(coin));
+        }
+    }
+}
+
+void CheckCache(const CBlock& block, const CCoinsViewCache& cache)
+{
+    uint32_t counter{0};
+    std::unordered_set<Txid, SaltedTxidHasher> txids{};
+    txids.reserve(block.vtx.size() - 1);
+
+    for (const auto& tx : block.vtx) {
+        if (tx->IsCoinBase()) {
+            BOOST_CHECK(!cache.GetPossiblySpentCoinFromCache(tx->vin[0].prevout));
+        } else {
+            for (const auto& in : tx->vin) {
+                const auto& outpoint{in.prevout};
+                const auto should_have{!txids.contains(outpoint.hash)};
+                if (should_have) {
+                    cache.AccessCoin(outpoint);
+                    ++counter;
+                }
+                const auto have{cache.GetPossiblySpentCoinFromCache(outpoint)};
+                BOOST_CHECK(should_have ? !!have : !have);
+            }
+            txids.emplace(tx->GetHash());
+        }
+    }
+    BOOST_CHECK(cache.GetCacheSize() == counter);
+}
+
+
+BOOST_FIXTURE_TEST_CASE(fetch_inputs_from_db, CoinsViewCacheAsyncTest)
+{
+    const auto& block{getBlock()};
+    NoAccessCoinsView dummy;
+    CCoinsViewCache db(&dummy);
+    PopulateCache(block, db);
+    CCoinsViewCache main_cache(&dummy);
+    CoinsViewCacheAsync view{main_cache, db};
+    for (auto i{0}; i < 3; ++i) {
+        view.StartFetching(block);
+        CheckCache(block, view);
+        view.Reset();
+    }
+}
+
+BOOST_FIXTURE_TEST_CASE(fetch_inputs_from_cache, CoinsViewCacheAsyncTest)
+{
+    const auto& block{getBlock()};
+    NoAccessCoinsView dummy;
+    CCoinsViewCache db(&dummy);
+    CCoinsViewCache main_cache(&dummy);
+    PopulateCache(block, main_cache);
+    CoinsViewCacheAsync view{main_cache, db};
+    for (auto i{0}; i < 3; ++i) {
+        view.StartFetching(block);
+        CheckCache(block, view);
+        view.Reset();
+    }
+}
+
+// Test for the case where a block spends coins that are spent in the cache, but
+// the spentness has not been flushed to the db.
+BOOST_FIXTURE_TEST_CASE(fetch_no_double_spend, CoinsViewCacheAsyncTest)
+{
+    const auto& block{getBlock()};
+    NoAccessCoinsView dummy;
+    CCoinsViewCache db(&dummy);
+    PopulateCache(block, db);
+    CCoinsViewCache main_cache(&dummy);
+    // Add all inputs as spent already in cache
+    PopulateCache(block, main_cache, /*spent=*/true);
+    CoinsViewCacheAsync view{main_cache, db};
+    for (auto i{0}; i < 3; ++i) {
+        view.StartFetching(block);
+        for (const auto& tx : block.vtx) {
+            for (const auto& in : tx->vin) view.AccessCoin(in.prevout);
+        }
+        // Coins are not added to the view, even though they exist unspent in the parent db
+        BOOST_CHECK(view.GetCacheSize() == 0);
```

Comment:
any reason not to use:
```suggestion
        BOOST_CHECK_EQUAL(view.GetCacheSize(), 0);
```
?